connmgr.go

```diff
@@ -164,8 +164,12 @@ func (cm *BasicConnMgr) TagPeer(p peer.ID, tag string, val int) {
 
 	pi, ok := cm.peers[p]
 	if !ok {
-		log.Info("tried to tag conn from untracked peer: ", p)
-		return
+		pi = &peerInfo{
```

Comment:
So, unfortunately, this will leak memory if we:

1. Connect.
2. Disconnect.
3. Tag the peer.

A real fix will, unfortunately, require access to a Host and/or Network object (to check if we're currently connected. Unfortunately... that requires a much larger refactor (which is why I haven't bothered fixing this...).

Reply:
> But as long as Connect and Disconnect are paired, the calls from a rational caller to TagPeer and UntagPeer are paired. So maybe the situation is improved a little by this patch.

Not necessarily. Peers are *sometimes* tagged within Connect/Disconnect hooks but many services tag peers on an as-needed basis (and may never explicitly untag them). The expectation is that the connection manager will clean up any tags after a final disconnect.